File: Scripts/film/fonts.py
import os
import platform
from typing import Optional, List, Dict
from dataclasses import dataclass

@dataclass(frozen=True)
class FontFile:
    name: str   # Human-readable font name (e.g., 'MicrosoftYaHei')
    path: str   # Full path to the .ttf/.ttc/.otf file

    def escape_path(self) -> str:

        if not self.path:
            return ''
        return self.path.replace('\\', '/').replace(':', '\\:').replace('"', '\\"')

    def __repr__(self):
        return f'FontFile(name="{self.name}", path="{self.path}")'

    def exists(self) -> bool:
        return os.path.exists(self.path)

class Fonts:
    WindowsFontPath = os.path.expandvars(r'%WINDIR%\Fonts')

    _font_extensions = {'.ttf', '.otf', '.ttc'}
    _font_map: Dict[str, FontFile] = {}
    _initialized = False

    _predefined_fonts = {
        # Windows
        'Arial': os.path.join(WindowsFontPath, 'arial.ttf'),
        'MicrosoftYaHei': os.path.join(WindowsFontPath, 'msyh.ttc'),
        'SimHei': os.path.join(WindowsFontPath, 'simhei.ttf'),
        'SimSun': os.path.join(WindowsFontPath, 'simsun.ttc'),
        'KaiTi': os.path.join(WindowsFontPath, 'simkai.ttf'),
        'FangSong': os.path.join(WindowsFontPath, 'simfang.ttf'),
        'TimesNewRoman': os.path.join(WindowsFontPath, 'times.ttf'),

        # macOS
        'PingFang': '/System/Library/Fonts/PingFang.ttc',
        'Heiti SC': '/System/Library/Fonts/STHeiti Light.ttc',

        # Linux / Cross-platform
        'DejaVuSans': '/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf',
        'DejaVuSans-Bold': '/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf',
        'NotoSansCJK': '/usr/share/fonts/opentype/noto/NotoSansCJK-Regular.ttc',
        'WenQuanYiZenHei': '/usr/share/fonts/wenquanyi/wqy-zenhei.ttc',
    }

    @classmethod
    def _get_default_font_dirs(cls) -> List[str]:
        system = platform.system()
        if system == 'Windows':
            return [cls.WindowsFontPath]
        elif system == 'Darwin':
            return [
                '/System/Library/Fonts', '/Library/Fonts', os.path.expanduser('~/Library/Fonts')
            ]
        elif system == 'Linux':
            return [
                '/usr/share/fonts', '/usr/local/share/fonts', os.path.expanduser('~/.fonts')
            ]
        return []

    @classmethod
    def _initialize(cls):
        '''Scan font directories and populate the font map.'''
        if cls._initialized:
            return

        for font_dir in cls._get_default_font_dirs():
            if not os.path.isdir(font_dir):
                continue

            for root, _, files in os.walk(font_dir):
                for file in files:
                    ext = os.path.splitext(file)[1].lower()
                    if ext in cls._font_extensions:
                        font_file = FontFile(name=os.path.splitext(file)[0], path=os.path.join(root, file))
                        if font_file.name not in cls._font_map:  # First match wins
                            cls._font_map[font_file.name] = font_file

        # Add predefined fonts
        for name, path in cls._predefined_fonts.items():
            if os.path.exists(path):
                cls._font_map[name] = FontFile(name=name, path=path)


        for name, font in cls._font_map.items():
            # print(f'Found font: {name} at {font.path}')
            setattr(cls, name, font)
        cls._initialized = True

    @classmethod
    def resolve(cls, name: str) -> Optional[FontFile]:
        '''Get full path to font file for a given name.'''
        cls._initialize()

        return cls._font_map.get(name)

    @classmethod
    def register(cls, name: str, path: str):

        cls._font_map[name] = FontFile(name=name, path=path)

    @classmethod
    def unregister(cls, name: str):
        cls._font_map.pop(name, None)

    @classmethod
    def list(cls) -> List[str]:
        cls._initialize()
        return sorted(cls._font_map.keys())

    @classmethod
    def dump(cls) -> Dict[str, str]:
        cls._initialize()
        return dict(cls._font_map)

    @classmethod
    def __getattr__(cls, name: str) -> FontFile:
        cls._initialize()
        if name in cls._font_map:
            return cls._font_map[name]
        raise AttributeError(f'Fonts has no font named "{name}"')

File: Scripts/film-cs/Fonts.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;

namespace FilmCS
{
    public class FontFile
    {
        public string Name { get; set; }
        public string Path { get; set; }
        public FontFile(string name, string path)
        {
            Name = name;
            Path = path;
        }
        public string EscapePath()
        {
            if (string.IsNullOrEmpty(Path)) return "";
            return Path.Replace("\\", "/").Replace(":", "\\:").Replace("\"", "\\\"");
        }
        public override string ToString() => $"FontFile(name=\"{Name}\", path=\"{Path}\")";
        public bool Exists() => File.Exists(Path);
    }

    public static class Fonts
    {
        private static readonly string WindowsFontPath = Environment.ExpandEnvironmentVariables("%WINDIR%\\Fonts");
        private static readonly HashSet<string> FontExtensions = new HashSet<string> { ".ttf", ".otf", ".ttc" };
        private static readonly Dictionary<string, FontFile> FontMap = new Dictionary<string, FontFile>();
        private static bool _initialized = false;
        private static readonly Dictionary<string, string> PredefinedFonts = new Dictionary<string, string>
        {
            { "Arial", Path.Combine(WindowsFontPath, "arial.ttf") },
            { "MicrosoftYaHei", Path.Combine(WindowsFontPath, "msyh.ttc") },
            { "SimHei", Path.Combine(WindowsFontPath, "simhei.ttf") },
            { "SimSun", Path.Combine(WindowsFontPath, "simsun.ttc") },
            { "KaiTi", Path.Combine(WindowsFontPath, "simkai.ttf") },
            { "FangSong", Path.Combine(WindowsFontPath, "simfang.ttf") },
            { "TimesNewRoman", Path.Combine(WindowsFontPath, "times.ttf") },
            { "PingFang", "/System/Library/Fonts/PingFang.ttc" },
            { "Heiti SC", "/System/Library/Fonts/STHeiti Light.ttc" },
            { "DejaVuSans", "/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf" },
            { "DejaVuSans-Bold", "/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf" },
            { "NotoSansCJK", "/usr/share/fonts/opentype/noto/NotoSansCJK-Regular.ttc" },
            { "WenQuanYiZenHei", "/usr/share/fonts/wenquanyi/wqy-zenhei.ttc" }
        };

        private static List<string> GetDefaultFontDirs()
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                return new List<string> { WindowsFontPath };
            if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
                return new List<string> { "/System/Library/Fonts", "/Library/Fonts", Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), "Library/Fonts") };
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
                return new List<string> { "/usr/share/fonts", "/usr/local/share/fonts", Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".fonts") };
            return new List<string>();
        }

        private static void Initialize()
        {
            if (_initialized) return;
            foreach (var fontDir in GetDefaultFontDirs())
            {
                if (!Directory.Exists(fontDir)) continue;
                foreach (var file in Directory.EnumerateFiles(fontDir, "*.*", SearchOption.AllDirectories))
                {
                    var ext = Path.GetExtension(file).ToLower();
                    if (FontExtensions.Contains(ext))
                    {
                        var fontFile = new FontFile(Path.GetFileNameWithoutExtension(file), file);
                        if (!FontMap.ContainsKey(fontFile.Name))
                            FontMap[fontFile.Name] = fontFile;
                    }
                }
            }
            foreach (var kv in PredefinedFonts)
            {
                if (File.Exists(kv.Value))
                    FontMap[kv.Key] = new FontFile(kv.Key, kv.Value);
            }
            _initialized = true;
        }

        public static FontFile Resolve(string name)
        {
            Initialize();
            return FontMap.ContainsKey(name) ? FontMap[name] : null;
        }
        public static void Register(string name, string path)
        {
            FontMap[name] = new FontFile(name, path);
        }
        public static void Unregister(string name)
        {
            if (FontMap.ContainsKey(name))
                FontMap.Remove(name);
        }
        public static List<string> List()
        {
            Initialize();
            return FontMap.Keys.OrderBy(x => x).ToList();
        }
        public static Dictionary<string, FontFile> Dump()
        {
            Initialize();
            return new Dictionary<string, FontFile>(FontMap);
        }
    }
}
